source/common/http/conn_manager_impl.h

```diff
@@ -209,6 +210,41 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void encodeData(Buffer::Instance& data, bool end_stream) override;
     void encodeTrailers(ResponseTrailerMap& trailers) override;
     void encodeMetadata(MetadataMapVector& metadata) override;
+    void setRequestTrailers(Http::RequestTrailerMapPtr&& request_trailers) override {
+      ASSERT(!request_trailers_);
+      request_trailers_ = std::move(request_trailers);
+    }
+    void setContinueHeaders(Http::ResponseHeaderMapPtr&& continue_headers) override {
+      ASSERT(!continue_headers_);
+      continue_headers_ = std::move(continue_headers);
+    }
+    void setResponseHeaders(Http::ResponseHeaderMapPtr&& response_headers) override {
+      // We'll overwrite the headers in the case where we fail the stream after upstream headers
+      // have begun filter processing but before they have been sent downstream.
+      response_headers_ = std::move(response_headers);
+    }
+    void setResponseTrailers(Http::ResponseTrailerMapPtr&& response_trailers) override {
+      ASSERT(!response_trailers_);
+      response_trailers_ = std::move(response_trailers);
+    }
+
+    // TODO(snowp): Create shared OptRef/OptConstRef helpers
+    Http::RequestHeaderMapOptRef requestHeaders() override {
```

Comment:
I'm going to go against the Envoy grain and say I find pointers much easier to wrap my brain around.
Feel free to leave it as is as I think Matt leans the other way.

Reply:
Using pointers end up being a bit more ergonomic, though the OptRef pattern is a bit clearer semantically, so not sure what the best solution is here. The OptRef will bring with it a lot of refactoring throughout the code base to avoid clunky conversion from OptRef to pointer, so it does require some work to make this change

Reply:
Feel free to revert back if you prefer, it's a stylistic preference based on making much more likely to have an errant deletion, though in practice it would be hard for that to happen.

Reply:
I think I'm okay with the opt ref approach, I'll follow up with refactors to clean things up